src/compiler/checker.ts

```diff
@@ -13033,8 +13033,16 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return length(target.typeParameters) === length(typeArguments) ? createTypeReference(target, concatenate(typeArguments, [thisArgument || target.thisType!])) : type;
         }
         else if (type.flags & TypeFlags.Intersection) {
-            const types = sameMap((type as IntersectionType).types, t => getTypeWithThisArgument(t, thisArgument, needApparentType));
-            return types !== (type as IntersectionType).types ? getIntersectionType(types) : type;
+            const intersectionType = type as IntersectionType;
+            const types = sameMap(intersectionType.types, t => getTypeWithThisArgument(t, thisArgument, needApparentType));
+            if (types === intersectionType.types) {
+                return type;
+            }
+            const result = getIntersectionType(types);
+            if (result.flags & TypeFlags.Intersection) {
+                (result as IntersectionType).withThisArgumentTarget = intersectionType;

```

Comment:
The problem that it solves is that [`getTypeWithThisArgument` call](https://github.dev/microsoft/TypeScript/blob/efc9c065a2caa52c5bebd08d730eed508075a78a/src/compiler/checker.ts#L13075-L13076) for the `Bar1` creates a **new** intersection type, one that doesn't have any reference (through `.target` or similar) to the "original" intersection (the base type).

So `createUnionOrIntersectionProperty` creates 2 separate synthetic properties for `a` for both the base type and the derived type (even though they both refer to the same symbols, declarations, etc). That, in turn, fails [`derived === base` check](https://github.dev/microsoft/TypeScript/blob/efc9c065a2caa52c5bebd08d730eed508075a78a/src/compiler/checker.ts#L44697-L44700) that is responsible for checking if the property was implemented or not.

This introduces some kind of a link between those types so I can trace this information back through the symbol links.

It's not a super elegant solution but it's the best I was able to figure out. I've tried to cache those property symbols and reuse them between both of those types but that led to a lot of problems. An alternative solution would be to make the mentioned `derived === base` check smarter but I also failed to figure out how to accurately compare 2 symbols like that.